### ISSUE TITLE
Allow the Deferred returned by Agent.request to be cancelled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4330]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4330|
|Type|enhancement|
|Created|2010-03-01 02:00:00Z|
|Branch|https://github.com/twisted/twisted/tree/webclient-request-cancel-4330-5|

Cancelling the `Deferred` returned by `Agent.request` should fail the `Deferred` with a `CancelledError` and either:

  1. Cancel the connection setup attempt, if it is still in progress.
  1. Disconnect abort the in-progress request, if the connection has been set up and a request has been initiated.

This used to depend on [#4329](https://github.com/twisted/twisted/issues/4329) to handle the former case; these days it depends on endpoints supporting cancellation.

<details><summary>Searchable metadata</summary>

```
trac-id__4330 4330
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_webclient_request_cancel_4330_5 branches/webclient-request-cancel-4330-5
branch_author__itamarst__exarkun__habnabit itamarst, exarkun, habnabit
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1267408800000000 1267408800000000
changetime__1370087057000000 1370087057000000
version__None None
owner__exarkun exarkun
cc__forsberg cc__dangra cc__chris-
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

[#4329](https://github.com/twisted/twisted/issues/4329) may be a perfectly reasonable way to handle this, but would depending on [#11002](https://github.com/twisted/twisted/issues/11002) instead be a reasonable option, or would that involve rewriting too much code?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Well, I basically finished [#4329](https://github.com/twisted/twisted/issues/4329) already.  And I like small changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [28848]) Branching to 'webclient-request-cancel-4330'

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj commented|
|-|-|

A discussion on IRC brought up the point that Agent now uses endpoints instead of `ClientCreator`, which are not yet cancelable.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [jonathanj](#note_5):
> A discussion on IRC brought up the point that Agent now uses endpoints instead of `ClientCreator`, which are not yet cancelable.

What?  Yes they are: http://twistedmatrix.com/documents/current/core/howto/endpoints.html#auto4

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Unfortunately [#5485](https://github.com/twisted/twisted/issues/5485)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_7):
> Unfortunately [#5485](https://github.com/twisted/twisted/issues/5485)

Ouch.  Well, thanks for the reference.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_9"></a>|@habnabit commented|
|-|-|

(In [34323]) Branching to 'webclient-request-cancel-4330-2'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_10"></a>|@habnabit commented|
|-|-|

(In [34324]) Initial implementation (with tests!) of Agent request cancellation.

refs [#4330](https://github.com/twisted/twisted/issues/4330)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set owner to @habnabit|
|-|-|

Thanks for implementing this! We should have more cancellable Deferreds. This is looking pretty good. However, some improvements are needed:

  1. Please write tests for cancelling after the response download has started, but hasn't finished. E.g. your download is too slow so you decide to stop it. To make sure you cover all cases maybe mention the covered state machine states in the cancellation tests, and make sure they match the full set of tests.
  1. Needs a news file.
  1. You should add a mention and/or example to the web client howto about the fact that you can cancel requests.
  1. If request was cancelled, and any bytes at all have been written (or for completeness probably always is easier), rather than using loseConnection, abortConnection should be used. Given that you may be dealing with uncooperative server, and/or you're giving up and don't care about writing any more, having the connection close immediately is what you want, as opposed to waiting until all write buffers are cleared, which may take an arbitrarily long amount of time.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

habnabit pointed out, re point 1, that there is no Deferred at that point in the code. So I'm not sure what I was referring to.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_17"></a>|@habnabit commented|
|-|-|

(In [36263]) Branching to 'webclient-request-cancel-4330-3'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_18"></a>|@habnabit commented|
|-|-|

(In [36264]) Merging forward (refs [#4330](https://github.com/twisted/twisted/issues/4330))

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve removed owner|
|-|-|

I've handle the review comments, build results here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/webclient-request-cancel-4330-3

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @itamarst<br>@itamarst set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst commented|
|-|-|

Thanks for the update!

 1. Cancelling when in TRANSMITTING eventually results in `loseConnection()`; this should really be `abortConnection()`. I believe the relevant code is `ebRequestWriting` in `HTTP11ClientProtocol.request`.
 2. For completeness sake it might be worth testing that cancellation works during the connection process, i.e. that `HTTPConnectionPool.getConnection` results are cancellable... but that just devolves to endpoint cancellation. So not sure quite what to do here. Probably nothing.
 3. The auto-retry code in `_RetryingHTTP11ClientProtocol._shouldRetry` should be modified so it does not retry if the reason for failure was cancellation.
 4. File a ticket for making the result of `FileBodyProducer.startProducing` cancellable.

I will address these comments and resubmit for review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_23"></a>|@itamarst commented|
|-|-|

(In [38499]) If an error occurs when generating a request body (e.g. cancellation), close the connection with abortConnection() rather than loseConnection().

Refs [#4330](https://github.com/twisted/twisted/issues/4330)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_24"></a>|@itamarst commented|
|-|-|

(In [38500]) Branching to 'webclient-request-cancel-4330-4'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_25"></a>|@itamarst commented|
|-|-|

(In [38502]) Don't retry requests that failed due to user cancellation.

Refs [#4330](https://github.com/twisted/twisted/issues/4330)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_26"></a>|@itamarst removed owner<br>@itamarst set status to `new`|
|-|-|

1.Done.
 2. Skipped.
 3. Done. Only tested one exception since it's the only one that can be the result of cancellation and cause a retry at the same time.
 4. Filed [#6531](https://github.com/twisted/twisted/issues/6531).

Started a new buildbot run, but have not yet looked at results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/webclient-request-cancel-4330-4

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks!  A few simple final comments:

  1. twisted/web/test/test_newclient.py
    1. `test_cancelDuringBodyProduction` calls `HTTP11ClientProtocol.connectionLost` - several of the other new cancel tests don't, though.  is there something special about this one?
    1. `test_cancelDuringChunkedBodyProduction` seems pretty similar to `test_cancelDuringBodyProduction`.  Maybe the repetition could be factored out into a helper.
    1. I agree that cancellation before connection setup is basically just endpoint cancellation.  Maybe there should be a test for `HTTPConnectionPool.getConnection` that exercises this cancellation code path, though?  Seems like it should be easy to write.  If you get stuck, I wouldn't be opposed to seeing this done in a follow-up ticket.
  1. twisted/web/test/test_agent.py
    1. I see the test docstring follows the local convention of using "should", but maybe fixing the existing docstrings is simple enough that we can make it all follow the modern style instead?
    1. `test_onlyRetryIdempotentMethods` and `test_onlyRetryIfNoResponseReceived` should be split up into multiple tests.  Can you file a ticket for that?
  1. twisted/web/client.py
    1. The change to `_shouldRetry` is incompletely tested.  There is no test coverage for a `_WrapperException` without a `CancelledError`

That's all.  Please address to your satisfaction and then merge.  Thanks again to everyone who contributed!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_29"></a>|@itamarst commented|
|-|-|

1.Done.
 2. Done, and [#6533](https://github.com/twisted/twisted/issues/6533).
 3. Done.

Started another buildbot run, if all looks good will merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Builds look good, I'll go ahead and merge this.  Thanks everybody!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

(In [38596]) Branching to 'webclient-request-cancel-4330-5'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

(In [38598]) put the news fragment in the right place

refs [#4330](https://github.com/twisted/twisted/issues/4330)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun set status to `closed`|
|-|-|

(In [38599]) Merge webclient-request-cancel-4330-5

Author: habnabit, therve, itamar
Reviewer: itamar, exarkun
Fixes: [#4330](https://github.com/twisted/twisted/issues/4330)

Implement resource cleanup for cancellation of the Deferred returned by
`twisted.web.client.Agent.request`.
